include/zephyr/zbus/zbus.h

```diff
@@ -0,0 +1,560 @@
+ * @retval -EINVAL Some parameter is invalid.
+ */
+int zbus_chan_notify(struct zbus_channel *chan, k_timeout_t timeout);
+
+#if defined(CONFIG_ZBUS_CHANNEL_NAME) || defined(__DOXYGEN__)
+/**
+ * @brief Get the channel's name.
+ *
+ * This routine returns the channel's name reference.
+ *
+ * @param chan The channel's reference.
+ *
+ * @return Channel's name reference.
+ */
+const char *zbus_chan_name(const struct zbus_channel *chan);
+#endif
+
+/**
+ * @brief Get the reference for a channel message directly.
+ *
+ * This routine returns the reference of a channel message.
+ *
+ * @warning This function cannot used directly for read-only channels.
+ * @warning This function must only be used directly for acquired (locked by mutex) channels. This
+ * can be done inside a listener for the receiving channel or after claim a channel.
+ *
+ * @param chan The channel's reference.
+ *
+ * @return Channel's message reference.
+ */
+void *zbus_chan_msg(struct zbus_channel *chan);
+
+/**
+ * @brief Get a constant reference for a channel message directly.
+ *
+ * This routine returns a constant reference of a channel message. This should be used
+ * inside listeners to access the message directly. In this way zbus prevents the listener of
+ * changing the notifying channel's message during the notification process.
+ *
+ * @note This function can be used directly for read-only channels.
+ * @warning This function must only be used directly for acquired (locked by mutex) channels. This
+ * can be done inside a listener for the receiving channel or after claim a channel.
+ *
+ * @param chan The channel's constant reference.
+ *
+ * @return A constant channel's message reference.
+ */
+void *zbus_chan_const_msg(const struct zbus_channel *chan);
+
+/**
+ * @brief Get the channel's message size.
+ *
+ * This routine returns the channel's message size.
+ *
+ * @param chan The channel's reference.
+ *
+ * @return Channel's message size.
+ */
+uint16_t zbus_chan_msg_size(const struct zbus_channel *chan);
+
+#if (CONFIG_ZBUS_CHANNEL_USER_DATA_SIZE > 0) || defined(__DOXYGEN__)
+/**
+ * @brief Get the channel's user data.
+ *
+ * This routine returns the channel's user data.
+ *
+ * @param chan The channel's reference.
+ *
+ * @return Channel's user data.
+ */
+void *zbus_chan_user_data(const struct zbus_channel *chan);
+#endif
+
+#if (CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE > 0) || defined(__DOXYGEN__)
+/**
+ * @brief Add an observer to a channel.
+ *
+ * This routine adds an observer to the channel.
+ *
+ * @param chan The channel's reference.
+ * @param obs The observer's reference to be added.
+ * @param timeout Waiting period to add an observer,
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval 0 Observer added to the channel.
+ * @retval -EINVAL Some parameter is invalid.
+ */
+int zbus_chan_add_obs(struct zbus_channel *chan, struct zbus_observer *obs, k_timeout_t timeout);
```

Comment:
Why use the observers pool? Why not make the runtime observer provide the memory for `struct zbus_observer_node`, for example in a macro like `ZBUS_RUNTIME_SUBSCRIBER_DEFINE()`?

Reply:
@desowin ping

Reply:
Why would you want to add the same observer to multiple channels? Shouldn't there be a 1:1 relationship between observer and a channel?

Reply:
Not at all. The relationship can be 1:1, 1:N, and N:N between publishers and observers. You can see the same observer added to several channels. It would be painful to have an observer for each channel we would like to observe. It would increase the code size unnecessarily.